ignore generated (expanded macros) code

#559

in https://gitlab.com/fommil/drone-dynamic-agents/ uncomment the Disable rule for `scala.Any.asInstanceOf` in `project/scalafix.conf` and do an `sbt lint`, you will see hundreds and hundreds of lines of output like

```
src/main/scala/interpreters/gcefs2.scala:81:6: error: [Disable.asInstanceOf] asInstanceOf is disabled and it got inferred as `*(io.circe.Decoder.importedDecoder[Cluster](new io.circe.export.Exported(((shapeless.lazily.apply[DerivedDecoder[Cluster]]({
  val inst$macro$347: DerivedDecoder[Cluster] = {
    final class anon$exportDecoder$macro$346 extends _root_.scala.Serializable {
      lazy val inst$macro$172: ReprDecoder[HNil] = io.circe.generic.decoding.ReprDecoder.hnilReprDecoder.asInstanceOf[ReprDecoder[HNil]];
      lazy val inst$macro$169: DerivedDecoder[STOPPING.type] = io.circe.generic.decoding.DerivedDecoder.deriveDecoder[STOPPING.type, hnilZipWithKeys.Out](shapeless.LabelledGeneric.materializeProduct[STOPPING.type, HNil, HNil, hnilZipWithKeys.Out]((({

...
```

--- a/build.sbt
+++ b/build.sbt
@@ -145,6 +145,7 @@ lazy val testsInput = project
     scalacOptions += warnAdaptedArgs.value, // For NoAutoTupling
     scalacOptions += warnUnusedImports.value, // For RemoveUnused
     scalacOptions += "-Ywarn-unused", // For RemoveUnusedTerms
+    scalacOptions ++= maxwarns.value, // Increase the maximum warnings to print
     logLevel := Level.Error, // avoid flood of compiler warnings
     libraryDependencies += bijectionCore,
     testsInputOutputSetting,

--- a/project/ScalafixBuild.scala
+++ b/project/ScalafixBuild.scala
@@ -39,9 +39,14 @@ object ScalafixBuild extends AutoPlugin with GhpagesKeys {
       crossVersion := CrossVersion.full
     )
     lazy val isScala213 = Def.setting { scalaVersion.value.startsWith("2.13") }
+    lazy val isScala212 = Def.setting { scalaVersion.value.startsWith("2.12") }
     lazy val warnUnusedImports = Def.setting {
       if (isScala213.value) "-Wunused:imports"
       else "-Ywarn-unused-import"
+    }
+    lazy val maxwarns = Def.setting {
+      if (isScala213.value || isScala212.value) Seq("-Xmaxwarns", "1000")
+      else Nil
     }
     lazy val scaladocOptions = Seq(
       "-groups",

--- a/scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala
+++ b/scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala
@@ -53,7 +53,7 @@ class RemoveUnused(config: RemoveUnusedConfig)
     val isUnusedImport = mutable.Set.empty[Position]
     val isUnusedPattern = mutable.Set.empty[Position]
 
-    val unusedPatterExpr = raw"^pattern var .* in [a-z]+ <.+> is never used".r
+    val unusedPatterExpr = raw"^pattern var .* in (value|method) .* is never used".r
 
     doc.diagnostics.foreach { diagnostic =>
       if (config.imports && diagnostic.message == "Unused import") {

--- a/scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala
+++ b/scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala
@@ -53,7 +53,8 @@ class RemoveUnused(config: RemoveUnusedConfig)
     val isUnusedImport = mutable.Set.empty[Position]
     val isUnusedPattern = mutable.Set.empty[Position]
 
-    val unusedPatterExpr = raw"^pattern var .* in (value|method) .* is never used".r
+    val unusedPatterExpr =
+      raw"^pattern var .* in (value|method) .* is never used".r
 
     doc.diagnostics.foreach { diagnostic =>
       if (config.imports && diagnostic.message == "Unused import") {

--- a/scalafix-tests/input/src/main/scala-2.13/test/RemoveUnusedPatternVars.scala
+++ b/scalafix-tests/input/src/main/scala-2.13/test/RemoveUnusedPatternVars.scala
@@ -8,7 +8,7 @@ case class AB(a: Int, b: String)
 case class XY(x: Int, y: Int)
 case class YZ(xy: XY, z: String)
 
-object Unusedmatchargs {
+object ob {
   // Add code that needs fixing here.
   val example = AB(42, "lol")
 
@@ -55,5 +55,321 @@ object Unusedmatchargs {
           println(b)
         }
       }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+}
+
+trait tr {
+  // Add code that needs fixing here.
+  val example = AB(42, "lol")
+
+  example match {
+    case AB(_, _) => println("Not used, good")
+  }
+  example match {
+    case AB(a, b) => println("Not used, wrong")
+  }
+  example match {
+    case AB(_, b) => println("b is not used, wrong")
+  }
+  example match {
+    case AB(a, _) => println("a is not used, wrong")
+  }
+
+  example match {
+    case AB(a, b) => println(s"$a $b used, good")
+  }
+  example match {
+    case AB(_, b) => println(s"$b is used, good")
+  }
+  example match {
+    case AB(a, _) => println(s"$a is used, good")
+  }
+
+  example match {
+    case AB(a, _) if a < Int.MaxValue =>
+      println("Do not delete")
+  }
+
+  example match {
+    case AB(a, _) =>
+      val a = 5
+      println(a)
+  }
+
+  val anotherExample = YZ(XY(1, 2), "3")
+
+  anotherExample match {
+    case YZ(el, b) =>
+      el match {
+        case XY(b, _) => {
+          println(b)
+        }
+      }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+}
+
+class cl {
+  // Add code that needs fixing here.
+  val example = AB(42, "lol")
+
+  example match {
+    case AB(_, _) => println("Not used, good")
+  }
+  example match {
+    case AB(a, b) => println("Not used, wrong")
+  }
+  example match {
+    case AB(_, b) => println("b is not used, wrong")
+  }
+  example match {
+    case AB(a, _) => println("a is not used, wrong")
+  }
+
+  example match {
+    case AB(a, b) => println(s"$a $b used, good")
+  }
+  example match {
+    case AB(_, b) => println(s"$b is used, good")
+  }
+  example match {
+    case AB(a, _) => println(s"$a is used, good")
+  }
+
+  example match {
+    case AB(a, _) if a < Int.MaxValue =>
+      println("Do not delete")
+  }
+
+  example match {
+    case AB(a, _) =>
+      val a = 5
+      println(a)
+  }
+
+  val anotherExample = YZ(XY(1, 2), "3")
+
+  anotherExample match {
+    case YZ(el, b) =>
+      el match {
+        case XY(b, _) => {
+          println(b)
+        }
+      }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(a, b) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, b) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(a, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(a, _) =>
+        val a = 5
+        println(a)
+    }
   }
 }

--- a/scalafix-tests/output/src/main/scala-2.13/test/RemoveUnusedPatternVars.scala
+++ b/scalafix-tests/output/src/main/scala-2.13/test/RemoveUnusedPatternVars.scala
@@ -5,7 +5,7 @@ case class AB(a: Int, b: String)
 case class XY(x: Int, y: Int)
 case class YZ(xy: XY, z: String)
 
-object Unusedmatchargs {
+object ob {
   // Add code that needs fixing here.
   val example = AB(42, "lol")
 
@@ -52,5 +52,321 @@ object Unusedmatchargs {
           println(b)
         }
       }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+}
+
+trait tr {
+  // Add code that needs fixing here.
+  val example = AB(42, "lol")
+
+  example match {
+    case AB(_, _) => println("Not used, good")
+  }
+  example match {
+    case AB(_, _) => println("Not used, wrong")
+  }
+  example match {
+    case AB(_, _) => println("b is not used, wrong")
+  }
+  example match {
+    case AB(_, _) => println("a is not used, wrong")
+  }
+
+  example match {
+    case AB(a, b) => println(s"$a $b used, good")
+  }
+  example match {
+    case AB(_, b) => println(s"$b is used, good")
+  }
+  example match {
+    case AB(a, _) => println(s"$a is used, good")
+  }
+
+  example match {
+    case AB(a, _) if a < Int.MaxValue =>
+      println("Do not delete")
+  }
+
+  example match {
+    case AB(_, _) =>
+      val a = 5
+      println(a)
+  }
+
+  val anotherExample = YZ(XY(1, 2), "3")
+
+  anotherExample match {
+    case YZ(el, _) =>
+      el match {
+        case XY(b, _) => {
+          println(b)
+        }
+      }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+}
+
+class cl {
+  // Add code that needs fixing here.
+  val example = AB(42, "lol")
+
+  example match {
+    case AB(_, _) => println("Not used, good")
+  }
+  example match {
+    case AB(_, _) => println("Not used, wrong")
+  }
+  example match {
+    case AB(_, _) => println("b is not used, wrong")
+  }
+  example match {
+    case AB(_, _) => println("a is not used, wrong")
+  }
+
+  example match {
+    case AB(a, b) => println(s"$a $b used, good")
+  }
+  example match {
+    case AB(_, b) => println(s"$b is used, good")
+  }
+  example match {
+    case AB(a, _) => println(s"$a is used, good")
+  }
+
+  example match {
+    case AB(a, _) if a < Int.MaxValue =>
+      println("Do not delete")
+  }
+
+  example match {
+    case AB(_, _) =>
+      val a = 5
+      println(a)
+  }
+
+  val anotherExample = YZ(XY(1, 2), "3")
+
+  anotherExample match {
+    case YZ(el, _) =>
+      el match {
+        case XY(b, _) => {
+          println(b)
+        }
+      }
+  }
+
+  def a: Unit = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
+  }
+
+  val b = {
+    example match {
+      case AB(_, _) => println("Not used, good")
+    }
+    example match {
+      case AB(_, _) => println("Not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("b is not used, wrong")
+    }
+    example match {
+      case AB(_, _) => println("a is not used, wrong")
+    }
+
+    example match {
+      case AB(a, b) => println(s"$a $b used, good")
+    }
+    example match {
+      case AB(_, b) => println(s"$b is used, good")
+    }
+    example match {
+      case AB(a, _) => println(s"$a is used, good")
+    }
+
+    example match {
+      case AB(a, _) if a < Int.MaxValue =>
+        println("Do not delete")
+    }
+
+    example match {
+      case AB(_, _) =>
+        val a = 5
+        println(a)
+    }
   }
 }

